cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,750 @@
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+// duration components timeparts structure
+struct alignas(4) duration_component {
+  int32_t day;        //-2,147,483,648 to 2,147,483,647
+  int32_t subsecond;  // 000000000 to 999999999
+  int8_t hour;        // 00 to 23
+  int8_t minute;      // 00 to 59
+  int8_t second;      // 00 to 59
+  bool is_negative;   // true/false
+};
+
+enum class format_char_type : int8_t {
+  literal,   // literal char type passed through
+  specifier  // duration format specifier
+};
+
+/**
+ * @brief Represents a format specifier or literal from a duration format string.
+ *
+ * Created by the format_compiler when parsing a format string.
+ */
+struct alignas(4) format_item {
+  format_char_type item_type;  // specifier or literal indicator
+  char value;                  // specifier or literal value
+  int8_t length;               // item length in bytes
+
+  static format_item new_specifier(char format_char, int8_t length)
+  {
+    return format_item{format_char_type::specifier, format_char, length};
+  }
+  static format_item new_delimiter(char literal)
+  {
+    return format_item{format_char_type::literal, literal, 1};
+  }
+};
+
+/**
+ * @brief The format_compiler parses a duration format string into a vector of
+ * format_items.
+ *
+ * The vector of format_items are used when parsing a string into duration
+ * components and when formatting a string from duration components.
+ */
+struct format_compiler {
+  std::string format;
+  rmm::device_uvector<format_item> d_items;
+
+  std::map<char, int8_t> specifier_lengths = {
+    {'-', -1},  // '-' if negative
+    {'D', -1},  // 1 to 11 (not in std::format)
+    {'H', 2},   // HH
+    {'I', 2},   // HH
+    {'M', 2},   // MM
+    {'S', -1},  // 2 to 13 SS[.mmm][uuu][nnn] (uuu,nnn are not in std::format)
+    {'p', 2},   // AM/PM
+    {'R', 5},   // 5 HH:MM
+    {'T', 8},   // 8 HH:MM:SS"
+    {'r', 11}   // HH:MM:SS AM/PM
+  };
+  format_compiler(const char* format_, cudaStream_t stream) : format(format_), d_items(0, stream)
+  {
+    std::vector<format_item> items;
+    const char* str = format.c_str();
+    auto length     = format.length();
+    bool negative_sign{true};
+    while (length > 0) {
+      char ch = *str++;
+      length--;
+      if (ch != '%') {
+        items.push_back(format_item::new_delimiter(ch));
+        continue;
+      }
+      CUDF_EXPECTS(length > 0, "Unfinished specifier in duration format");
+
+      ch = *str++;
+      length--;
+      if (ch == '%')  // escaped % char
+      {
+        items.push_back(format_item::new_delimiter(ch));
+        continue;
+      } else if (ch == 'n') {
+        items.push_back(format_item::new_delimiter('\n'));
+        continue;
+      } else if (ch == 't') {
+        items.push_back(format_item::new_delimiter('\t'));
+        continue;
+      }
+      if (ch == 'O') {
+        CUDF_EXPECTS(*str == 'H' || *str == 'I' || *str == 'M' || *str == 'S',
```

Comment:
What does `%O` do?  I do not see it mentioned in the doxygen in for the API.
Also, there should be a gtest for this error case.